File: packages/nodejs/src/domains/auth/jwt.ts
import http from 'http';
import { JwtPayload, sign, verify, decode } from 'jsonwebtoken';
import {
  AUTH_SCHEME,
  DEFAULT_JWT_EXPIRATION_SEC,
  JWT_HASH_ALGORITHM,
} from '../../constants';
import { jwtUninitialized } from '../../errors';
import { getDebug } from '../../logging';
import { isSignedout } from './signout';

const debug = getDebug('domains:auth:jwt');

export type ProviderFunction = (req: http.IncomingMessage) => Promise<{
  issuer: string;
  audience: string[];
}>;

export interface JwtConfig {
  secret: string;
  provider: string | ProviderFunction;
  expirationSec?: number;
}

export interface JwtClaims {
  exp: number;
  iat: number;
  nbf: number;
  sub: string;
  iss: string;
  aud: string | string[];
}

const regExpAuthScheme = new RegExp(`^${AUTH_SCHEME}$`, 'i');

class Jwt {
  secret!: string;
  provider!: string | ProviderFunction;
  expirationSec!: number;

  constructor(config: JwtConfig) {
    this.secret = config.secret;
    this.provider = config.provider;
    this.expirationSec = config.expirationSec ?? DEFAULT_JWT_EXPIRATION_SEC;
  }

  private async getIssuerAndAudience(req: http.IncomingMessage): Promise<{
    issuer: string;
    audience: string[];
  }> {
    if (typeof this.provider === 'string') {
      return {
        issuer: this.provider,
        audience: [this.provider],
      };
    }
    return this.provider(req);
  }

  // 署名
  async sign(subject: string, req: http.IncomingMessage): Promise<string> {
    const now = Math.floor(Date.now() / 1000);
    const { issuer, audience } = await this.getIssuerAndAudience(req);
    const token = sign(
      {
        exp: now + this.expirationSec, // 有効期限
        iat: now, // 発行した日時
        nbf: 0, // 有効になる日時
        sub: subject, // ユーザー識別子
        iss: issuer, // 発行者
        aud: audience, // 利用者
      },
      this.secret,
      {
        algorithm: JWT_HASH_ALGORITHM,
      }
    );
    return `${AUTH_SCHEME} ${token}`;
  }

  // 検証
  async verify(
    token: string,
    req: http.IncomingMessage
  ): Promise<JwtClaims | null> {
    const [scheme, credentials, ...unexpects] = token.split(' ');
    if (unexpects?.length || !regExpAuthScheme.test(scheme)) {
      return null;
    }
    const { issuer, audience } = await this.getIssuerAndAudience(req);
    return await new Promise((resolve): void => {
      verify(
        credentials,
        this.secret,
        {
          algorithms: [JWT_HASH_ALGORITHM],
          audience,
          issuer,
          complete: false,
        },
        (err: Error | null, decoded?: JwtPayload | string): void => {
          if (err) {
            debug(err);
            return resolve(null);
          }
          debug('decoded claims:', decoded);
          resolve(decoded ? (decoded as JwtClaims) : null);
        }
      );
    });
  }

  // デコード
  decode(token: string): JwtClaims | null {
    const [scheme, credentials, ...unexpects] = token.split(' ');
    if (unexpects?.length || !regExpAuthScheme.test(scheme)) {
      return null;
    }
    return decode(credentials) as JwtClaims;
  }
}

let jwt: Jwt | undefined;

// 初期化
export const initJwt = (config: JwtConfig, force = false): Jwt => {
  if (!jwt || force) {
    jwt = new Jwt(config);
  }
  return jwt;
};

// JWT生成
export const signJwt = async (
  subject: string,
  req: http.IncomingMessage
): Promise<string> => {
  if (!jwt) {
    throw jwtUninitialized();
  }
  return await jwt.sign(subject, req);
};

// JWT検証
export const verifyJwt = async (
  token: string,
  req: http.IncomingMessage
): Promise<JwtClaims | null> => {
  if (!jwt) {
    throw jwtUninitialized();
  }
  if (await isSignedout(token)) {
    debug('Already signed out. token: %s', token);
    return null;
  }
  return await jwt.verify(token, req);
};

// JWTデコード
export const decodeJwt = async (token: string): Promise<JwtClaims | null> => {
  if (!jwt) {
    throw jwtUninitialized();
  }
  return jwt.decode(token);
};

File: packages/golang/domains/auth/jwt.go
package auth

import (
	"context"
	"fmt"
	"net/http"
	"time"

	"github.com/cam-inc/viron/packages/golang/logging"

	"github.com/cam-inc/viron/packages/golang/errors"

	"github.com/cam-inc/viron/packages/golang/constant"

	"github.com/cam-inc/viron/packages/golang/config"

	"github.com/lestrrat-go/jwx/jwa"
	jwxJwt "github.com/lestrrat-go/jwx/jwt"

	"github.com/go-chi/jwtauth"
)

type (
	// Config struct {
	// 	Secret        string
	// 	Provider      string
	// 	ExpirationSec int
	// }
	Claim struct {
		Exp int
		Iat int
		Nbf int
		Sub string
		Iss string
		Aud []string
	}
)

var (
	jwt *config.JWT
	log logging.Logger
)

func SetUp(secret string, provider func(r *http.Request) (string, []string, error), expiration int) error {
	jwt = &config.JWT{
		Secret:        secret,
		Provider:      provider,
		ExpirationSec: expiration,
		JwtAuth:       jwtauth.New(string(jwa.HS512), []byte(secret), nil),
	}
	log = logging.GetDefaultLogger()
	return nil
}

func Sign(r *http.Request, subject string) (string, error) {
	iss, aud, err := jwt.Provider(r)
	if err != nil {
		return "", err
	}
	claim := map[string]interface{}{
		"nbf": 0,
		"sub": subject,
		"iss": iss,
		"aud": aud,
	}
	jwtauth.SetExpiryIn(claim, time.Duration(jwt.ExpirationSec)*time.Second)
	jwtauth.SetIssuedNow(claim)
	_, tokenStr, err := jwt.JwtAuth.Encode(claim)

	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%s %s", constant.AUTH_SCHEME, tokenStr), nil
}

func Verify(r *http.Request, token string) (*Claim, error) {

	if jwt == nil {
		return nil, errors.JwtUninitialized
	}

	if IsSignedOut(context.Background(), token) {
		return nil, fmt.Errorf("this token is revoked %s", token)
	}

	jwtToken, err := VerifyToken(token)
	if err != nil {
		return nil, err
	}

	iss, aud, err := jwt.Provider(r)
	if err != nil {
		return nil, err
	}
	if iss != jwtToken.Issuer() {
		return nil, fmt.Errorf("iss miss match verify iss[%s] jwt iss[%s]", iss, jwtToken.Issuer())
	}
	if !checkAudience(aud, jwtToken.Audience()) {
		return nil, fmt.Errorf("aud miss match verify aud[%s] jwt aud[%s]", aud, jwtToken.Audience())
	}

	claim := &Claim{
		Exp: int(jwtToken.Expiration().Unix()),
		Iat: int(jwtToken.IssuedAt().Unix()),
		Nbf: int(jwtToken.NotBefore().Unix()),
		Sub: jwtToken.Subject(),
		Iss: jwtToken.Issuer(),
		Aud: jwtToken.Audience(),
	}

	return claim, nil
}

func VerifyToken(token string) (jwxJwt.Token, error) {
	return jwtauth.VerifyToken(jwt.JwtAuth, token)
}

func checkAudience(providerAudience []string, jwtAudience []string) bool {
	// 発行者と依頼者が未設定はtrue
	if len(providerAudience) == 0 && len(jwtAudience) == 0 {
		return true
	}
	// jwt audience が provider audienceに含まれていればtrue
	for _, ja := range jwtAudience {
		for _, pa := range providerAudience {
			if ja == pa {
				return true
			}
		}
	}
	return false
}
